amundsen_application/static/js/components/common/ResourceList/PaginatedResourceList/index.spec.tsx

```diff
@@ -0,0 +1,119 @@
+
+import { ResourceType } from 'interfaces';
+import ResourceListItem from 'components/common/ResourceListItem/index';
+
+import PaginatedResourceList, { PaginatedResourceListProps } from './';
+import * as CONSTANTS from '../constants';
+
+describe('PaginatedResourceList', () => {
+  const setStateSpy = jest.spyOn(PaginatedResourceList.prototype, 'setState');
+  const setup = (propOverrides?: Partial<PaginatedResourceListProps>) => {
+    const props: PaginatedResourceListProps = {
+      allItems: [
+        { type: ResourceType.table },
+        { type: ResourceType.table },
+        { type: ResourceType.table },
+        { type: ResourceType.table },
+        { type: ResourceType.table },
+        { type: ResourceType.table },
+      ],
+      itemsPerPage: 4,
+      source: 'testSource',
+      ...propOverrides
+    };
+    const wrapper = shallow<PaginatedResourceList>(<PaginatedResourceList {...props} />);
+    return { props, wrapper };
+  };
+
+  describe('componentDidUpdate', () => {
+    it('updates activePage state if the activePage is out of bounds by multiple pages', () => {
+      const wrapper = setup({ itemsPerPage: 2 }).wrapper;
+      wrapper.setState({ activePage: 2 });
+      const componentDidUpdateSpy = jest.spyOn(wrapper.instance(), 'componentDidUpdate');
+      wrapper.setProps({
+        allItems: [
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+        ]
+      });
+      expect(componentDidUpdateSpy).toHaveBeenCalled();
+      expect(wrapper.state().activePage).toEqual(0);
+    })
+
+    it('updates activePage state if the activePage is out of bounds by one page', () => {
+      const wrapper = setup({ itemsPerPage: 2 }).wrapper;
+      wrapper.setState({ activePage: 2 });
+      const componentDidUpdateSpy = jest.spyOn(wrapper.instance(), 'componentDidUpdate');
+      wrapper.setProps({
+        allItems: [
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+        ]
+      })
+      expect(componentDidUpdateSpy).toHaveBeenCalled();
+      expect(wrapper.state().activePage).toEqual(1);
+    })
+
+    it('does not update activePage if new values are not out of bounds', () => {
+      const wrapper = setup({ itemsPerPage: 2 }).wrapper;
+      wrapper.setState({ activePage: 2 });
+      wrapper.setProps({
+        allItems: [
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+          { type: ResourceType.table },
+        ]
+      })
+      expect(wrapper.state().activePage).toEqual(2);
+    })
+  });
+
+  describe('onPagination', () => {
+    it('calls setState to update the activePage', () => {
+      setStateSpy.mockClear();
+      const wrapper = setup().wrapper;
+      wrapper.instance().onPagination(3);
+      expect(setStateSpy).toHaveBeenCalledWith({ activePage: 2 });
+    });
+  });
```

Comment:
IMHO this test goes too much into the details.

We could do something similar by rendering the component, calling onPagination with the value 3, and then seeing how the rendered component changes.

Reply:
Would the suggestion be correctly summarized as "unit test `onPagination` by verifying its effects on the rendered result"?

If so, one of the things Daniel and I chose to move toward was the concept of writing tests that would not break if and when we made changes in the logic of other methods. This exact case is not the worst kind of case since te rendering for this component is basically just affected by a function of `activePage` and `itemsPerPage`, and `itemsPerPage` is always a constant. However for more complicated components I would see us more likely to be on the hook for modifying assertions in these methods if we get new designs or otherwise tweak the UI as we iterate through features. I'm open to talking more about unit test approach pros and cons though

Reply:
Yes, you word it better than I did. The public effects of clicking on "onPagination" are seen on the rendering of the component (and by extension, of the Pagination component), while calling 'setState' can be considered an implementation detail.

You are right that in this case, checking those public effects would mean that we will search on the rendered result of the Pagination component, making this more like a small "integration test". Maybe a middle ground would be to check the props of the Pagination component after rendering it, something like:
`wrapper.find(Pagination).props().activePage`. I think this is less problematic than checking the state or spying on setState.

Reply:
As mentioned above, let's make a goal to do some knowledge share on these topics!